vpr/src/pack/sync_netlists_to_routing_flat.cpp

```diff
@@ -0,0 +1,367 @@
+
+        for(int pin: pins_to_erase){
+            pb_routes.erase(pin);
+        }
+    }
+
+    /* Go through each route tree and rebuild the pb_routes */
+    for(ParentNetId net_id: atom_ctx.nlist.nets()){
+        auto& tree = route_ctx.route_trees[net_id];
+        if(!tree)
+            continue; /* No routing at this ParentNetId */
+
+        /* Get all intrablock connections */
+        std::vector<std::pair<RRNodeId, RRNodeId>> conns_to_restore; /* (source, sink) */
+        get_intra_cluster_connections(tree.value(), conns_to_restore);
+
+        /* Restore the connections */
+        for(auto [source_inode, sink_inode]: conns_to_restore){
+            auto physical_tile = device_ctx.grid.get_physical_type({
+                rr_graph.node_xlow(source_inode),
+                rr_graph.node_ylow(source_inode),
+                rr_graph.node_layer(source_inode)
+            });
+            int source_pin = rr_graph.node_pin_num(source_inode);
+            int sink_pin = rr_graph.node_pin_num(sink_inode);
+
+            auto [_, subtile] = get_sub_tile_from_pin_physical_num(physical_tile, source_pin);
+
+            ClusterBlockId clb = placement_ctx.grid_blocks.block_at_location({
+                rr_graph.node_xlow(source_inode),
+                rr_graph.node_ylow(source_inode),
+                subtile,
+                rr_graph.node_layer(source_inode)
+            });
+
+            /* Look up pb graph pins from pb type if pin is not on tile, look up from block otherwise */
+            const t_pb_graph_pin* source_pb_graph_pin, *sink_pb_graph_pin;
+            if(is_pin_on_tile(physical_tile, sink_pin)){
+                sink_pb_graph_pin = get_pb_graph_node_pin_from_block_pin(clb, sink_pin);
+            }else{
+                sink_pb_graph_pin = get_pb_pin_from_pin_physical_num(physical_tile, sink_pin);
+            }
+            if(is_pin_on_tile(physical_tile, source_pin)){
+                source_pb_graph_pin = get_pb_graph_node_pin_from_block_pin(clb, source_pin);
+            }else{
+                source_pb_graph_pin = get_pb_pin_from_pin_physical_num(physical_tile, source_pin);
+            }
+
+            t_pb* pb = cluster_ctx.clb_nlist.block_pb(clb);
+
+            /* Route between the pins */
+            route_intra_cluster_conn(source_pb_graph_pin, sink_pb_graph_pin, convert_to_atom_net_id(net_id), pb);
+        }
+    }
+}
+
+/** Rebuild the ClusterNetId <-> AtomNetId lookup after compressing the ClusterNetlist
+ * Needs the "most recent" ClusterNetIds in atom_ctx.lookup: won't work after invalidating the ClusterNetIds twice */
+inline void rebuild_atom_nets_lookup(ClusteredNetlist::IdRemapper& remapped){
+    auto& atom_ctx = g_vpr_ctx.mutable_atom();
+    auto& atom_lookup = atom_ctx.lookup;
+
+    for(auto parent_net_id: atom_ctx.nlist.nets()){
+        auto atom_net_id = convert_to_atom_net_id(parent_net_id);
+        ClusterNetId old_clb_net = atom_lookup.clb_net(atom_net_id);
+        if(!old_clb_net)
```

Comment:
That works because we update the atom_clb_net lookup while rebuilding the netlist: https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/2691/files#diff-a94e475b89606b112d5a6d4ef22d94b1db792391571fc1d9545687466913ade0R310-R313